A bunch of bug fixes and some enhancements

Fesaa opened the pull request:
Kavita will now try using all volumes/chapters (specials, prologue, etc) when deciding if a series has completed. This may cause some false positives. But should reduce false positives in the other direction drastically. 

# Changed
- Changed: Changed how Kavita+ decides if a series is completed (Closes #3794)
- Changed: Known For section on a person page will now take all roles into account (Closes #3858)
- Changed: Kavita+ will now sync scrobble at a random minute instead of at the start of the hour

# Fixed
- Fixed: Fixed browse Genres/Tags missing from smaller devices (develop) (Fixes #3856)
- Fixed: Fixed exception while trying to assign average rating when Kavita+ returned no ratings (Fixes #3843)
- Fixed: Fixed Kavita+ publication status being removed after the series is scanned, Kavita+ now locks the field (Fixed #3827)
- Fixed: Fixed Kavita+ publication status being wrong for EPUBs (Fixes #3794)
- Fixed: Fixed Kavita+ publication status using volume count when series do not have any volumes (Fixes #3794)
- Fixed: Fixed destination values autofilling the first letter of the source value (Genre/Tag mappings)
- Fixed: Fixed the Known For filter sometimes not returning any results (Fixes #3858)
- Fixed: Fixed Genres, Tags not being restricted to Library (Fixes #3859) (develop)
- Fixed: Fixed People not being age, or library restricted (Fixes #3822)
- Fixed: Fixed scrobble history on mobile devices not being readable (develop)

Fesaa committed:
Construct mobile menu and nav items dropdown from the same list #3856

Fesaa committed:
Fix exception while setting Chapter Ratings when no reviews are found fixes #3843

Fesaa committed:
Lock publication status after setting it with K+ fixes #3827

Fesaa committed:
Loosen requirements for Publication status fixes #3794

- Don't use volume count for MaxCount if there are no volumes
- Try using specials to reach external volume count
- Try matching chapters loosly (>=) when matching external chapter count
- Try using specials (prologue's, extra's) when matching external chapter count

Fesaa committed:
Don't auto fill destination values

Instead, filter out empty destination values when updating. It ends up just filling in the first letter of the source, which is just awkward to work with

Fesaa committed:
Fix empty filters after clicking on know for, don't filter on roles for known for Fixes #3858

Fesaa committed:
Restrict Genres and Tags to the users library Fixes #3859

Fesaa committed:
Restrict Person by library & Age rating

Fesaa committed:
Unit tests for ChapterTags

Fesaa committed:
Missed a file

Fesaa committed:
Merge branch 'develop' into bugfix/cleanup

Fesaa committed:
Genre & Person Unit Tests

Fesaa committed:
Merge branch 'develop' into bugfix/cleanup

Fesaa committed:
Fix mobile view scrobble history

Fesaa committed:
Schedule Scrobble sync at a random hour

Fesaa committed:
Get Person Count working correctly
Fix up some unit tests

Fesaa committed:
Less DB trips

majora2007 committed:
Don't use static variables to allow tests to pass

Fesaa committed:
Remove unused code, always restrict against age rating

Make comments more clear

Fesaa committed:
Include some unit tests for the new K+ Completed check

majora2007 commented on API.Tests/Repository/TagRepositoryTests.cs:
```diff
@@ -0,0 +1,205 @@
+
+public class TagRepositoryTests : AbstractDbTest
+{
+    private static readonly Tag SharedSeriesChaptersTag = new TagBuilder("Shared Series Chapter Tag").Build();
```
We need to refactor this like I did with Person repo, so we don't use static fields.

majora2007 commented on API/Data/Repositories/GenreRepository.cs:
```diff
@@ -173,23 +173,56 @@ public async Task<PagedList<BrowseGenreDto>> GetBrowseableGenre(int userId, User
+            query = query.Where(s => s.Chapters.Any(cp => seriesIds.Contains(cp.Volume.SeriesId)) ||
+                                     s.SeriesMetadatas.Any(sm => seriesIds.Contains(sm.SeriesId)));
+
+            finalQuery = query.Select(g => new BrowseGenreDto
```
The final query is the same, so just modify the query with the if, then do the select, so there is no code duplication.

Fesaa replied:
There is a slight different in the count sub queries. One also filters on library, while the other doesn't. I've tried making this one and using WhereIf, but then I get those errors:
```
System.InvalidOperationException
The LINQ expression 'sm => DbSet<Series>()
    .Where(s3 => __8__locals1_userLibs_0
        .Contains(s3.LibraryId))
    .Select(s3 => s3.Id)
    .Contains(sm.SeriesId)' could not be translated. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'. See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
```

majora2007 replied:
I understand, WhereIf works on a queryable, but SeriesMetadatas is ICollection

majora2007 commented on API/Data/Repositories/PersonRepository.cs:
```diff
@@ -204,42 +209,82 @@ public async Task<PagedList<BrowsePersonDto>> GetBrowsePersonDtos(int userId, Br
+                CoverImage = p.CoverImage,
+                SeriesCount = p.SeriesMetadataPeople
+                    .Select(smp => smp.SeriesMetadata)
+                    .Where(sm => seriesIds.Contains(sm.SeriesId))
```
Use WhereIf(shouldLibRestrict, sm => seriesIds.Contains(sm.SeriesId) so we don't have to duplicate the query select?

Fesaa replied:
See the comment above, I've tried this but always get exceptions

majora2007 commented on API/Data/Repositories/PersonRepository.cs:
```diff
@@ -386,27 +439,31 @@ public async Task<IList<Person>> GetAllPeople(PersonIncludes includes = PersonIn
     }
 
-    public async Task<IList<PersonDto>> GetAllPersonDtosAsync(int userId, PersonIncludes includes = PersonIncludes.Aliases)
+    public async Task<IList<PersonDto>> GetAllPersonDtosAsync(int userId, PersonIncludes includes = PersonIncludes.None)
```
Why was this changed?

Fesaa replied:
The signature in the interface doesn't include it; the public endpoint where this method is called used for this does not require them either (used for creating filter options)

majora2007 commented on API/Data/Repositories/PersonRepository.cs:
```diff
@@ -386,27 +439,31 @@ public async Task<IList<Person>> GetAllPeople(PersonIncludes includes = PersonIn
     }
 
-    public async Task<IList<PersonDto>> GetAllPersonDtosByRoleAsync(int userId, PersonRole role, PersonIncludes includes = PersonIncludes.Aliases)
+    public async Task<IList<PersonDto>> GetAllPersonDtosByRoleAsync(int userId, PersonRole role, PersonIncludes includes = PersonIncludes.None)
```
Why was this changed?

Fesaa replied:
Same reason as above

majora2007 commented on API/Data/Repositories/TagRepository.cs:
```diff
@@ -111,24 +111,57 @@ public async Task<PagedList<BrowseTagDto>> GetBrowseableTag(int userId, UserPara
+            query = query.Where(tag => tag.Chapters.Any(cp => seriesIds.Contains(cp.Volume.SeriesId)) ||
+                                       tag.SeriesMetadatas.Any(sm => seriesIds.Contains(sm.SeriesId)));
+
+            finalQuery = query
```
See comments for other repos

majora2007 commented on API/Extensions/QueryExtensions/RestrictByLibraryExtensions.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace API.Extensions.QueryExtensions;
+
+// TODO: Refactor with IQueryable userLibs? But then I can't do the allLibs check?
```
What is the TODO?

Fesaa replied:
Forgot to remove, I did this already

majora2007 commented on API/Extensions/QueryExtensions/RestrictByLibraryExtensions.cs:
```diff
@@ -0,0 +1,36 @@
+
+// TODO: Refactor with IQueryable userLibs? But then I can't do the allLibs check?
+/// <summary>
+/// Optionally pass ids of all libraries, will then be smart and not restrict if the person has access to all
```
This summary doesn't make sense here. Looks like it needs to be deleted.

majora2007 commented on API/Services/Plus/ExternalMetadataService.cs:
```diff
@@ -1621,6 +1628,58 @@ private PublicationStatus DeterminePublicationStatus(Series series, List<Chapter
+        // Note: I've currently opted to keep this an equals to prevent the above bug from happening
+        // We *could* change this to >= in the future in case this is reported by users
+        // If we do; test IsSeriesCompleted_Volumes_TooManySpecials needs to be updated
+        if (maxVolumes != Parser.DefaultChapterNumber && series.Volumes.Count == externalMetadata.Volumes)
```
I think this should be + 1 if you have specials, because specials get put into their own volume. AL's volume count does not include specials.

Fesaa replied:
I'm actually not sure how AL does it, I've really only seen them add Specials to Chapter counts; and just assumed they might do the same for volumes. 

If they don't include specials with volumes like they do with chapters, might be better to just remove this case altogether?

majora2007 replied:
Validated this in our discord privately.

majora2007 reviewed (changes requested):
Overall looking good, a few things in terms of cleanup.

majora2007 committed:
Refactored to not use static variables.

majora2007 committed:
Converted genre repo tests to use non-static fields

majora2007 committed:
Switched to use a count query instead of serializing all libraries.

Fesaa committed:
Simply code, moving to my PC

majora2007 committed:
Small fix

majora2007 committed:
It aint pretty, but it works.

Fesaa committed:
Ensure we set the correct MaxCount & TotalCount when matching on modifiedMinVolumeCount

majora2007 commented on API.Tests/Services/ExternalMetadataServiceTests.cs:
```diff
@@ -928,6 +928,36 @@ public void IsSeriesCompleted_Volumes_IncludeSpecialsCheck()
+                .WithTotalCount(3)
+                .Build())
+            .WithVolume(new VolumeBuilder("1").WithNumber(1).Build())
+            .WithVolume(new VolumeBuilder("1.5").WithNumber(1.5f).Build())
```
This is not a special chapter, we need to either treat it as as special (specials go into VolumeBuilder(Parser.SpecialVolume)) or we need to update the wording. 

This is testing that partial chapters don't clog up the count.

Fesaa committed:
Get naming correctly in unit tests, add some notes

Adjust numbers in unit tests to make sure they're values that can be passed to IsSeriesCompleted (TotalCount = External count unless comicinfo has more)
